classes/data-stores/ActionScheduler_DBStore.php

```
@@ -375,56 +375,103 @@ protected function get_query_actions_sql( array $query, $select_or_count = 'sele
+		 ) );
 
 		/** @var \wpdb $wpdb */
 		global $wpdb;
-		$sql        = ( 'count' === $select_or_count ) ? 'SELECT count(a.action_id)' : 'SELECT a.action_id';
-		$sql       .= " FROM {$wpdb->actionscheduler_actions} a";
-		$sql_params = array();
+		$mysql_version = $wpdb->db_version();
+		$sql           = ( 'count' === $select_or_count ) ? 'SELECT count(a.action_id)' : 'SELECT a.action_id';
+		$sql           .= " FROM {$wpdb->actionscheduler_actions} a";
+		$sql_params    = array();
 
 		if ( ! empty( $query['group'] ) || 'group' === $query['orderby'] ) {
 			$sql .= " LEFT JOIN {$wpdb->actionscheduler_groups} g ON g.group_id=a.group_id";
 		}
 
-		$sql .= ' WHERE 1=1';
+		$sql .= " WHERE 1=1";
 
 		if ( ! empty( $query['group'] ) ) {
-			$sql         .= ' AND g.slug=%s';
+			$sql          .= " AND g.slug=%s";
 			$sql_params[] = $query['group'];
 		}
 
-		if ( $query['hook'] ) {
-			$sql         .= ' AND a.hook=%s';
+		if ( ! empty( $query['hook'] ) ) {
+			$sql          .= " AND a.hook=%s";
 			$sql_params[] = $query['hook'];
 		}
-		if ( ! is_null( $query['args'] ) ) {
-			$sql         .= ' AND a.args=%s';
-			$sql_params[] = $this->get_args_for_query( $query['args'] );
+
+		if ( ! is_null( $query['args'] ) && isset( $query['partial_args_matching'] ) ) {
+			switch ( $query['partial_args_matching'] ) {
+				case 'json':
+					if ( $mysql_version < '5.7' ) {
+						throw new \RuntimeException( __( 'JSON partial matching requires MySQL 5.7+.', 'action-scheduler' ) );
+					}
+					foreach ( $query['args'] as $key => $value ) {
+						$supported_types = array(
+							'integer' => '%d',
+							'boolean' => '%s',
+							'double'  => '%f',
+							'string'  => '%s',
+						);
+						foreach ( $supported_types as $type => $placeholder ) {
+							if ( gettype( $value ) === $type ) {
+								if ( 'boolean' === $type ) {
+									$value = $value ? 'true' : 'false';
+								}
+								$sql .= ' AND JSON_EXTRACT(a.args, %s)='.$placeholder;
+							}
+						}
```

Instead of iterating over all supported types, and given we already know the type, we could do something like the following, which might be a bit simpler:

```php
$value_type  = gettype( $value );
$placeholder = isset( $supported_types[ $value_type ] ) ? $supported_types[ $value_type ] : false;
...
```

That way, we can also react to types that aren't supported (say, "object" or "array"). Not sure what we want to do in those cases, though. Do we want to silently ignore the arg or throw an exception? What are your thoughts @alexmigf?

---

I think we should throw an exception, at least we give some feedback if something is not supported.